community/kernel/src/main/java/org/neo4j/kernel/api/schema/CompositeIndexDescriptor.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.schema;
+
+import org.neo4j.storageengine.api.schema.SchemaRule;
+
+import static java.lang.String.format;
+
+/**
+ * Description of a single index.
```

Comment:
Currently we require that a composite index has multiple keys. There is even a check or two that will throw an exception if this is not the case. It might not be the best model, but it is a good intermediate state where it forces test failures for situations where we mix them up. I think we could improve this further, and have ideas for that, but suggest we do it in later PRs.

An earlier version of this work back in December simply generalized all indexes to composite, and the old behaviour was implied by the property array having length 1. We moved away from that model because it was harder to be sure that code that expected single property indexes didn't inadvertently use only one property of the list. Hence the current model which is stricter. And I think while we can improve the current model, going back to the previous one does not feel like an improvement.